File: day52/majorityElement.java
/*
169.多数元素
给定一个大小为 n 的数组，找到其中的多数元素。多数元素是指在数组中出现次数大于 ⌊ n/2 ⌋ 的元素。
你可以假设数组是非空的，并且给定的数组总是存在多数元素。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/majority-element
 */
 
 class Solution {
    public int majorityElement(int[] nums) {
        Map<Integer, Integer> hashmap = new HashMap<>();
        for (int num : nums) {
            hashmap.put(num, hashmap.getOrDefault(num, 0) + 1);
        }

        Map.Entry<Integer, Integer> majorityEntry = null;
        for (Map.Entry<Integer, Integer> entry : hashmap.entrySet()) {
            if (majorityEntry == null || entry.getValue() > majorityEntry.getValue()) {
                majorityEntry = entry;
            }
        }
        
        return majorityEntry.getKey();
    }
}

// class Solution {
//     public int majorityElement(int[] nums) {
//         int count = 0, res = 0;
//         for (int num : nums) {
//             if (count == 0) res = num;
//             count += res == num ? 1 : -1;
//         }
//         return res;
//     }
// }

// class Solution {
//     public int majorityElement(int[] nums) {
//         Arrays.sort(nums);
//         return nums[nums.length / 2];
//     }
// }

File: day52/majorityElement.py
# 分治法
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        def conquer(left, right):
            if left == right - 1: return nums[left]

            mid = (right - left) // 2 + left
            win_left = conquer(left, mid)
            win_right = conquer(mid, right)
            if win_left == win_right: return win_left

            left_count = sum(1 for num in nums[left:right] if num == win_left)
            right_count = sum(1 for num in nums[left:right] if num == win_right)
            return win_left if left_count > right_count else win_right

        return conquer(0, len(nums))

# 摩尔法
# class Solution:
#     def majorityElement(self, nums: List[int]) -> int:
#         count = 0
#         res = None
#         for num in nums:
#             if count == 0: res = num
#             count += 1 if res == num else -1
#         return res

# 哈希映射法
# class Solution:
#     def majorityElement(self, nums: List[int]) -> int:
#         counts = collections.Counter(nums)
#         return max(counts.keys(), key=counts.get)

# class Solution:
#     def majorityElement(self, nums: List[int]) -> int:
#         dic = dict()
#         for num in nums:
#             dic[num] = dic.get(num, 0) + 1
#         return max(dic, key=lambda x: dic[x])

# 排序法
# class Solution:
#     def majorityElement(self, nums: List[int]) -> int:
#         nums.sort()
#         return nums[len(nums) // 2]
